File: scripts/hooks/comment_budget_guard.py
#!/usr/bin/env python3
"""PreToolUse hook (Edit|Write): backstop for the Comments rule in
rules/coding-standards.md. Python port of the home box's comment-budget-guard.js.

Denies the write before the file is touched when the new content has:
  - more than 3 comment lines in a row, or
  - more than 20% comment density, once the content is 15+ lines.

Skips markdown, JSON, config files, anything under docs/, Python docstrings
(only `#` lines count), and JSDoc blocks carrying @param / @returns.
Escape hatch: CLAUDE_COMMENT_BUDGET=off.

Contract: reads hook JSON on stdin, writes hook JSON on stdout. Fails open —
a broken guard must not break the session.
"""

from __future__ import annotations

import json
import os
import re
import sys

MAX_RUN = 3
MAX_DENSITY = 0.20
MIN_LINES_FOR_DENSITY = 15

SKIP_EXTENSIONS = {
    ".md", ".markdown", ".rst", ".txt", ".json", ".jsonl", ".yaml", ".yml",
    ".toml", ".ini", ".cfg", ".conf", ".env", ".csv", ".html", ".xml", ".lock",
}
HASH_COMMENT = {".py", ".sh", ".bash", ".ps1", ".r", ".rb", ".pl", ".cmake"}
SLASH_COMMENT = {
    ".js", ".jsx", ".ts", ".tsx", ".mjs", ".cjs", ".java", ".c", ".h", ".cpp",
    ".hpp", ".cs", ".go", ".rs", ".swift", ".kt", ".scala", ".css", ".scss",
}
DASH_COMMENT = {".sql"}


def comment_flags(lines: list[str], ext: str) -> list[bool]:
    flags = [False] * len(lines)
    if ext in HASH_COMMENT:
        for i, line in enumerate(lines):
            stripped = line.strip()
            flags[i] = stripped.startswith("#") and not stripped.startswith("#!")
    elif ext in DASH_COMMENT:
        for i, line in enumerate(lines):
            flags[i] = line.strip().startswith("--")
    elif ext in SLASH_COMMENT:
        in_block = False
        block_start = 0
        for i, line in enumerate(lines):
            stripped = line.strip()
            if in_block:
                flags[i] = True
                if "*/" in stripped:
                    in_block = False
                    if any("@param" in lines[j] or "@returns" in lines[j] for j in range(block_start, i + 1)):
                        for j in range(block_start, i + 1):
                            flags[j] = False
                continue
            if stripped.startswith("//"):
                flags[i] = True
            elif stripped.startswith("/*"):
                flags[i] = True
                block_start = i
                if "*/" in stripped:
                    if "@param" in stripped or "@returns" in stripped:
                        flags[i] = False
                else:
                    in_block = True
    return flags


def apply_edit(content: str, old: str, new: str, replace_all: bool) -> str | None:
    if not old or old not in content:
        return None
    return content.replace(old, new) if replace_all else content.replace(old, new, 1)


def header_end(lines: list[str], ext: str) -> int:
    flags = comment_flags(lines, ext)
    i = 0
    while i < len(lines):
        stripped = lines[i].strip()
        if stripped == "" or stripped.startswith("#!") or flags[i]:
            i += 1
            continue
        break
    return i


# A run survives blank lines so a block split by whitespace still counts as one.
def check_run(content: str, ext: str) -> str | None:
    lines = content.splitlines()
    flags = comment_flags(lines, ext)
    run = 0
    for i, flag in enumerate(flags):
        if flag:
            run += 1
        elif lines[i].strip() != "":
            run = 0
        if run > MAX_RUN:
            return (
                f"more than {MAX_RUN} comment lines in a row — the Comments rule "
                "(rules/coding-standards.md) allows one line, two at most, WHY not "
                "WHAT. Cut the block, then retry. Escape: CLAUDE_COMMENT_BUDGET=off."
            )
    return None


def check_density(file_text: str, ext: str) -> str | None:
    lines = file_text.splitlines()
    body = lines[header_end(lines, ext):]
    if not body:
        return None
    total = len(body)
    count = sum(comment_flags(body, ext))
    if total >= MIN_LINES_FOR_DENSITY and count / total > MAX_DENSITY:
        return (
            f"comment density {count}/{total} = {count / total:.0%} exceeds "
            f"{MAX_DENSITY:.0%} in the resulting file — the Comments rule "
            "(rules/coding-standards.md) says one line, two at most, WHY not WHAT. "
            "Trim comments, then retry. Escape: CLAUDE_COMMENT_BUDGET=off."
        )
    return None


sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "lib"))
try:
    from record_component import record
except Exception:  # the recorder is optional, the deny is not
    def record(**_: object) -> None:
        return None


def main() -> None:
    if os.environ.get("CLAUDE_COMMENT_BUDGET", "").lower() == "off":
        return
    payload = json.load(sys.stdin)
    tool_name = payload.get("tool_name")
    if tool_name not in ("Edit", "Write"):
        return
    tool_input = payload.get("tool_input") or {}
    file_path = (tool_input.get("file_path") or "").replace("\\", "/")
    ext = os.path.splitext(file_path)[1].lower()
    if ext in SKIP_EXTENSIONS or ext not in (HASH_COMMENT | SLASH_COMMENT | DASH_COMMENT):
        return
    if re.search(r"(^|/)docs(/|$)", file_path, re.IGNORECASE):
        return
    content = tool_input.get("content") or tool_input.get("new_string") or ""
    reason = check_run(content, ext)
    if not reason:
        # Density is judged on the file the edit produces, not the snippet alone.
        file_text = content if tool_name == "Write" else None
        if tool_name == "Edit":
            try:
                with open(tool_input.get("file_path") or "", "r", encoding="utf-8") as fh:
                    current = fh.read()
                file_text = apply_edit(current, tool_input.get("old_string") or "",
                                        tool_input.get("new_string") or "",
                                        tool_input.get("replace_all") is True)
            except Exception:
                file_text = None
        if file_text is not None:
            reason = check_density(file_text, ext)
    if reason:
        record(kind="hook", name=os.path.basename(__file__), session_id=payload.get("session_id"),
               cwd=payload.get("cwd"), blocked=True, notes=reason)
        print(json.dumps({
            "hookSpecificOutput": {
                "hookEventName": "PreToolUse",
                "permissionDecision": "deny",
                "permissionDecisionReason": reason,
            }
        }))


if __name__ == "__main__":
    try:
        main()
    except Exception:
        pass
    sys.exit(0)

File: scripts/hooks/comment-budget-guard.js
#!/usr/bin/env node
// Denies an Edit/Write whose comments exceed the budget in rules/coding-standards.md.
// Source files only. Set CLAUDE_COMMENT_BUDGET=off to disable.

const fs = require('fs');
const MAX_RUN = 3;
const MAX_DENSITY = 0.2;
const DENSITY_FLOOR = 15;
// @type must not match "@types/x" or "@typescript". @typedef and friends are API docs too.
const API_DOC = /@param|@returns|@return|@type(?![a-zA-Z])|@typedef|@property|@callback|@example|@throws/;

const LINE_TOKENS = {
  js: '//', jsx: '//', ts: '//', tsx: '//', mjs: '//', cjs: '//',
  go: '//', rs: '//', java: '//', c: '//', h: '//', cpp: '//', hpp: '//',
  cs: '//', swift: '//', kt: '//', scala: '//', php: '//', dart: '//',
  py: '#', sh: '#', bash: '#', zsh: '#', ps1: '#', rb: '#', pl: '#', r: '#',
  sql: '--', lua: '--',
};

const isDocQuote = (s) => s.startsWith('"""') || s.startsWith("'''");

// One pass labels every line, so run and density can never disagree.
// A block is 'doc' if any line in it carries an API tag, not just the opener.
function classify(lines, token) {
  const kind = new Array(lines.length).fill('code');
  let i = 0;
  while (i < lines.length) {
    const line = lines[i].trim();
    if (line === '') { kind[i] = 'blank'; i += 1; continue; }
    if (token === '#' && isDocQuote(line)) {
      const q = line.slice(0, 3);
      let j = i;
      if (!line.slice(3).includes(q)) { j += 1; while (j < lines.length && !lines[j].includes(q)) j += 1; }
      for (let k = i; k <= Math.min(j, lines.length - 1); k += 1) kind[k] = 'doc';
      i = j + 1; continue;
    }
    if (line.startsWith('/*')) {
      let j = i;
      while (j < lines.length && !lines[j].includes('*/')) j += 1;
      // A one-line /** */ is a field's doc comment, the TSDoc editors show on hover.
      // Narrative bloat is the multi-line untagged block, which still counts.
      const isApi = (j === i && line.startsWith('/**')) ||
        lines.slice(i, Math.min(j + 1, lines.length)).some((l) => API_DOC.test(l));
      for (let k = i; k <= Math.min(j, lines.length - 1); k += 1) kind[k] = isApi ? 'doc' : 'comment';
      i = j + 1; continue;
    }
    if (line.startsWith(token) && !line.startsWith('#!')) { kind[i] = 'comment'; i += 1; continue; }
    i += 1;
  }
  return kind;
}

// A run survives blank lines so a block split by whitespace still counts as one.
function longestRun(lines, token) {
  const kind = classify(lines, token);
  let best = 0;
  let run = 0;
  for (const k of kind) {
    if (k === 'comment') { run += 1; best = Math.max(best, run); continue; }
    if (k === 'blank') continue;
    run = 0;
  }
  return best;
}

function countComments(lines, token) {
  return classify(lines, token).filter((k) => k === 'comment').length;
}

// The single place that decides what is code, so no consumer can drift from the classifier.
function codeOnly(text, token) {
  const lines = text.split(/\r?\n/);
  const kind = classify(lines, token);
  return lines.filter((l, i) => kind[i] === 'code').map((l) => l.trimEnd()).join('\n');
}

// Exported so scan-comments.js grades files with this exact logic, never a copy.
module.exports = { LINE_TOKENS, MAX_RUN, MAX_DENSITY, DENSITY_FLOOR, classify, longestRun, countComments, codeOnly };
if (require.main !== module) return;

let record = () => {};
try { ({ record } = require('./lib/record-component')); } catch (e) { /* recorder missing: keep denying */ }

// Manual slicing, never String#replace(old, new_string): new_string may contain "$&"/"$1".
function applyEdit(content, oldStr, newStr, replaceAll) {
  if (!oldStr || content.indexOf(oldStr) === -1) return null;
  if (!replaceAll) {
    const idx = content.indexOf(oldStr);
    return content.slice(0, idx) + newStr + content.slice(idx + oldStr.length);
  }
  let result = '';
  let rest = content;
  let pos;
  while ((pos = rest.indexOf(oldStr)) !== -1) {
    result += rest.slice(0, pos) + newStr;
    rest = rest.slice(pos + oldStr.length);
  }
  return result + rest;
}

// Shebang/blank/comment header lines are boilerplate, not the body the density budget grades.
function headerEnd(lines, token) {
  const kind = classify(lines, token);
  let i = 0;
  while (i < lines.length) {
    const trimmed = lines[i].trim();
    if (trimmed === '' || trimmed.startsWith('#!') || kind[i] === 'comment') { i += 1; continue; }
    break;
  }
  return i;
}

function grade(fileText, token) {
  const lines = fileText.split(/\r?\n/);
  const body = lines.slice(headerEnd(lines, token));
  const comments = countComments(body, token);
  const density = body.length ? comments / body.length : 0;
  return { comments, total: body.length, density, over: body.length >= DENSITY_FLOOR && density > MAX_DENSITY };
}

let raw = '';
process.stdin.on('data', (d) => (raw += d));
process.stdin.on('end', () => {
  if (process.env.CLAUDE_COMMENT_BUDGET === 'off') process.exit(0);
  let input;
  try { input = JSON.parse(raw); } catch { process.exit(0); }
  if (input.tool_name !== 'Edit' && input.tool_name !== 'Write') process.exit(0);

  const ti = input.tool_input || {};
  const filePath = String(ti.file_path || '').split('\\').join('/');
  const ext = (filePath.split('.').pop() || '').toLowerCase();
  const token = LINE_TOKENS[ext];
  if (!token) process.exit(0);
  if (/\/(docs|node_modules|dist|build|vendor|\.venv)\//i.test(filePath)) process.exit(0);

  const text = String(ti.content != null ? ti.content : ti.new_string || '');
  if (!text) process.exit(0);
  const lines = text.split(/\r?\n/);
  const run = longestRun(lines, token);

  // Density is judged on the file the edit produces, not the snippet alone, so a
  // short why-comment dropped into a big file can't read as 100% comments.
  let fileText = input.tool_name === 'Write' ? text : null;
  let before = null;
  if (input.tool_name === 'Edit') {
    try {
      before = fs.readFileSync(ti.file_path, 'utf8');
      fileText = applyEdit(before, String(ti.old_string || ''), String(ti.new_string || ''), ti.replace_all === true);
    } catch (e) {
      fileText = null;
    }
  }

  const pct = (x) => Math.round(x * 100) + '%';
  let why = null;
  if (run > MAX_RUN) {
    why = run + ' comment lines in a row (budget: ' + MAX_RUN + ')';
  } else if (fileText != null) {
    const now = grade(fileText, token);
    const was = before != null ? grade(before, token) : null;
    if (was && was.over) {
      // Debt the file already carried is not this edit's to pay; what the edit adds must meet the budget.
      const addedComments = now.comments - was.comments;
      const addedLines = now.total - was.total;
      if (addedComments > 0 && addedComments > MAX_DENSITY * addedLines) {
        why = 'an edit adding ' + addedComments + ' comment lines in ' + addedLines + ' net new lines, on a file already at ' +
          pct(was.density) + ' (budget: ' + pct(MAX_DENSITY) + ')';
      }
    } else if (now.over) {
      why = now.comments + ' comment lines out of ' + now.total + ' in the resulting file (' +
        pct(now.density) + ', budget: ' + pct(MAX_DENSITY) + ')';
    }
  }
  if (!why) process.exit(0);

  const reason =
    'comment-budget-guard: ' + filePath + ' has ' + why + '. Comments are one line, two at most. ' +
    'Say WHY a non-obvious choice was made, never WHAT the code does. Measurement notes, dates and ' +
    'incident history go in docs/ARCHITECTURE.md or CLAUDE.md, never in source. Cut the block, then retry.';
  record({
    kind: 'hook',
    name: 'comment-budget-guard.js',
    sessionId: input.session_id,
    cwd: input.cwd,
    blocked: 1,
    notes: reason,
  });
  process.stdout.write(JSON.stringify({
    hookSpecificOutput: {
      hookEventName: 'PreToolUse',
      permissionDecision: 'deny',
      permissionDecisionReason: reason,
    },
  }));
  process.exit(0);
});
